### ISSUE TITLE
Cache hits "hide" warnings

### ISSUE THREAD

**Issue by rlepigre** [ocaml/dune#7460]
## Expected Behavior

Even when the dune cache is enabled, if I do `dune clean && dune build` I expect to see all warnings (and other output) produced by the commands that were (supposed to be) run. In other words, I kind of expect the output of commands to be cached as well.

## Actual Behavior

If the cache hits, nothing is printed, which makes it very easy to overlook warnings.

**Comment by rlepigre** [comment]
Another related thought I had on this issue: since the output of commands (which is not limited to warnings) may be non-deterministic (e.g., containing timing samples), I think it would be great if the cache supported a notion of "volatile" file that could be used for `stdout` and `stderr`, but not only. Ideally, we would be able to specify that the program being run, say `coqc`, will produce an extra output with extension `.perf`, and that such a file is to be cached and restored as-is, as long as the main artifacts of the compilation run are still the same.

This is something that would be invaluable for CI in Coq projects, where it is often valuable to generate performance profiles for individual `coqc` runs. At BedRock Systems, we currently have a hack to get this, which consists in encoding the performance data inside of Coq's `.glob` files (via a shim around the `coqc` command). This works great, but means that the `.glob` files are not deterministic, and this sometimes produces dune cache warnings. Ideally, we would not have to encode this data in `.glob` files, but annotate our Coq theory stanza with something like `(volatile %{basename}.perf)`, to let dune know that `coqc ... file.v` will not only produce `file.vo` and `file.glob`, but also a volatile file `file.perf` that is to be cached specially (together with the contents of `stdout` and `stderr`).

**Comment by ejgallego** [comment]
Hi @rlepigre , I guess the only way to deal with such files in the dune model is to have them depend on `(universe)`, as they will be re-run every time.

As far as I can tell, support for non-deterministic builds is not on the roadmap.

On the topic of build meta-data tho, many things can be done, see for example:

- https://github.com/ocaml/dune/issues/3449
- https://github.com/coq/coq/issues/11542

Also `fcc` in `coq-lsp` has built-in support for profiling, however that's not usable in your scope yet, but could be worth looking into it at some point.

**Comment by Alizter** [comment]
We've been thinking about changing warn-as-error back to warnings as the default. One of the problems we need to solve is the ability to replay warnings during a build. Such a feature would solve this issue most likely.

**Comment by rlepigre** [comment]
Thanks for your reply @ejgallego! For us it is really critical to use the cache, so depending on `(universe)` is a no go. Also, I'm not advocating for supporting non-deterministic builds, but for simply for allowing one to attach possibly non-deterministic data to deterministic builds. This seems quite reasonable to me, and potentially also easy to implement for someone familiar with the caching code.

@Alizter did you link the wrong issue?

**Comment by ejgallego** [comment]
> for allowing one to attach possibly non-deterministic data to deterministic builds

As of today we need to make Dune aware of this data, which is later used in the `--trace-file` option. If the data we want is performance data, it makes sense to me to reuse this data; once that is done, things should work fine with the cache too.

So maybe a first experiment would be to piggyback on the support for timing data Dune already has.

Do you expect other kind of "volatile" data beyond perf number / build meta-data?

**Comment by Alizter** [comment]
@rlepigre No that is the correct issue, it is just poorly named. Inside is the idea that `dune build` replays warnings which is a feature that should in theory work with the cache.

**Comment by rlepigre** [comment]
> So maybe a first experiment would be to piggyback on the support for timing data Dune already has.

That is indeed an option, but I'm not sure how well that would work. In particular, this is not very flexible, and requires outputting data in a format I cannot choose.
 
> Do you expect other kind of "volatile" data beyond perf number / build meta-data?

No, I can't think of anything else. But performance data can take many form, not all of which integrate well with the catapult trace format I think. For instance, you might want to gather data at several levels, in a way that they are not easily combined into a single file. For instance, you could imaging generating traces for Ltac2 functions gathered with some custom plugin, and also run `coqc` with `-profile`.

**Comment by rlepigre** [comment]
> @rlepigre No that is the correct issue, it is just poorly named. Inside is the idea that `dune build` replays warnings which is a feature that should in theory work with the cache.

Ah sorry, I did not read far enough. The first messages in the issue seem kind of unrelated, and then it diverges into basically a duplicate of the current issue. :smiley:

**Comment by 7h3kk1d** [comment]
There's some further discussion of this on discuss [here](https://discuss.ocaml.org/t/inconsistencies-in-dune-warnings/15730). Being able to run code/tests with warnings present is a huge boon but hiding warnings (especially in watch mode) breaks stuff for CLI-driven users.
